mongo/mongo.py

```diff
@@ -38,6 +42,17 @@ def backup(self):
         """
         LOG.info("starting mongo backup...")
 
+        mongo_uri = [uri for uri in self.extra if "--uri" in uri]
+        if(len(mongo_uri) > 0):
+            original_mongo_uri = mongo_uri[0]
+            mongo_uri_password = re.search(r"//([^:]+):(.*)@", mongo_uri[0]).group(2)
+            mongo_uri_username = re.search(r"//([^:]+):(.*)@", mongo_uri[0]).group(1)
+            mongo_uri_password_encoded = urllib.parse.quote(mongo_uri_password)
+            mongo_uri_username_encoded = urllib.parse.quote(mongo_uri_username)
+            original_mongo_uri = original_mongo_uri.replace(mongo_uri_password, mongo_uri_password_encoded)
+            original_mongo_uri = original_mongo_uri.replace(mongo_uri_username, mongo_uri_username_encoded)
+            self.extra = [original_mongo_uri]
```

Comment:
I like the use of urllib here, can we go further and use urllib to parse the mongo URI and modify the username/password instead of relying on regex?

Reply:
urllib only parses url's so I don't think we can use it to parse the username and password, but I will try to find a better way to do this. Also yes, that's an issue if the URI is already encoded then this will fail.

Reply:
> urllib only parses url's so I don't think we can use it to parse the username and password, but I will try to find a better way to do this. Also yes, that's an issue if the URI is already encoded then this will fail.

You can

```py
Python 3.9.1 (default, Dec 10 2020, 10:36:35)
[Clang 12.0.0 (clang-1200.0.32.27)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from urllib.parse import urlparse
>>> o = urlparse('mongodb://user:pass@mongo.localhost.com:5432')
>>> o
ParseResult(scheme='mongodb', netloc='user:pass@mongo.localhost.com:5432', path='', params='', query='', fragment='')
>>> o.username
'user'
>>> o.password
'pass'
>>>
```